device/transport/mqtt/src/mqtt_base.ts

```diff
@@ -54,167 +134,126 @@ export class MqttBase {
 
-    if (config.sharedAccessSignature) {
-      this._options.password = new Buffer(config.sharedAccessSignature.toString());
-      debug('username: ' + this._options.username);
+    if (this._config.sharedAccessSignature) {
+      options.password = this._config.sharedAccessSignature.toString();
+      debug('username: ' + options.username);
       debug('uri:      ' + uri);
     } else {
-      this._options.cert = config.x509.cert;
-      this._options.key = config.x509.key;
-      this._options.passphrase = config.x509.passphrase;
+      options.cert = this._config.x509.cert;
+      options.key = this._config.x509.key;
+      (<any>options).passphrase = this._config.x509.passphrase; // forced to cast to any because passphrase is used by tls options but not surfaced by the types definition.
     }
 
-    this.client = this.mqttprovider.connect(uri, this._options);
-    /*Codes_SRS_NODE_COMMON_MQTT_BASE_16_007: [The `connect` method shall not throw if the `done` argument has not been passed.]*/
-    if (done) {
-      const self = this;
-
-      const errCallback = (error) => {
+    const createErrorCallback = (eventName) => {
+      return (error) => {
+        debug('received \'' + eventName + '\' from mqtt client');
         const err = error || new errors.NotConnectedError('Unable to establish a connection');
-        self.client.removeListener('close', errCallback);
-        self.client.removeListener('offline', errCallback);
-        self.client.removeListener('disconnect', errCallback);
-        self.client.removeListener('error', errCallback);
-        done(err);
+        callback(err);
       };
+    };
+    /*Codes_SRS_NODE_COMMON_MQTT_BASE_16_003: [The `connect` method shall call the `done` callback with a standard javascript `Error` object if the connection failed.]*/
+    const errorCallback = createErrorCallback('error');
+    const closeCallback = createErrorCallback('close');
+    const offlineCallback = createErrorCallback('offline');
+    const disconnectCallback = createErrorCallback('disconnect');
+    const messageCallback = (topic, payload) => {
+      this.emit('message', topic, payload);
+    };
 
-      /*Codes_SRS_NODE_COMMON_MQTT_BASE_16_003: [The `connect` method shall call the `done` callback with a standard javascript `Error` object if the connection failed.]*/
-      self.client.on('error', errCallback);
-      self.client.on('close', errCallback);
-      self.client.on('offline', errCallback);
-      self.client.on('disconnect', errCallback);
-
-      self.client.on('connect', (connack) => {
-        debug('Device is connected');
-        debug('CONNACK: ' + JSON.stringify(connack));
+    this._mqttClient = this.mqttprovider.connect(uri, options);
+    this._mqttClient.on('message', messageCallback);
+    this._mqttClient.on('error', errorCallback);
+    this._mqttClient.on('close', closeCallback);
+    this._mqttClient.on('offline', offlineCallback);
+    this._mqttClient.on('disconnect', disconnectCallback);
```

Comment:
Shouldn't these be before the this.mqttprovider.connect call two lines above?

Reply:
`this._mqttClient` is the returned value of the `connect` call, so we can't subscribe before we have the object. we have to bank on the fact that connect returns synchronously in the underlying mqtt library.

---
In reply to: [140911193](https://github.com/Azure/azure-iot-sdk-node/pull/90#discussion_r140911193) [](ancestors = 140911193)